apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2466,25 +2486,108 @@ func (u *UniterAPI) GoalStates(args params.Entities) (string, error) {
 			result.Results[i].Error = common.ServerError(common.ErrPerm)
 			continue
 		}
-		// TODO (agprado): get units and relations information from
+		unit, err := u.getUnit(tag)
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		application, err := unit.Application()
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+		} else {
+			result.Results[i].Result, err = u.goalStateResult(application)
+			if err != nil {
+				result.Results[i].Error = common.ServerError(err)
+			}
+		}
+	}
+	return result, nil
+}
+
+// goalStateResult creates the structure with an application units and unit relations.
+func (u *UniterAPI) goalStateResult(application *state.Application) (*params.GoalState, error) {
+	gs := params.GoalState{}
+
+	var err error
+	gs.Units, err = u.goalStateUnits(application)
+	if err != nil {
+		return nil, err
+	}
+	allRelations, err := application.Relations()
+	if err != nil {
+		return nil, err
+	}
+	if allRelations != nil {
+		gs.Relations, err = u.goalStateRelations(allRelations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &gs, nil
```

Comment:
Why return a pointer?

Reply:
Generally, returning a pointer to a struct is useful since you can use nil if there's an error. Although in the params package we always seem to return an empty struct. This might be a requirement of the api marshalling, not sure. But for consistency at the very least, we should return a struct.